### ISSUE TITLE
Export a plot to a .pdf or .ps file. 

### ISSUE THREAD

**Issue by 3fon3fonov** [pyqtgraph/pyqtgraph#1455]
Please, please, please make this possible. Exporting to .png is usually fine, but this is not what we want to put in our scientific papers. The preferred plots must be of higher quality stored in a .pdf or .ps files.

Some time ago I open a question about this in the google groups, but only lately I got some suggestions.

https://groups.google.com/g/pyqtgraph/c/EnHt19ogGh0

Please look there for more details.

(SVG export simply does not work for me)

Cheers,
Trifon

**Comment by j9ac9k** [comment]
I'm not opposed to this feature, but I'm not sure any of the maintainers will be jumping on this.  There is a QPdfWriter but it is only a Qt5 feature, so we likely wouldn't merge this until after we've deprecated Qt4 support (which frankly isn't that far off).

I would advise trying to use the matplotlib exporter where there are more export options available; a better strantegy might be to identify why svg exports are not working for you.  About a month ago I merged #1401 which did some fixes to the SVG exporter (although there is still a known issue involving scatter plot items not showing up).

**Comment by 2xB** [comment]
I'll take this opportunity to suggest one thing that was certainly also discussed in the past: The more exporters we have, the more they can break and the more there is to maintain. Actually, having a rock-stable matplotlib exporter (maybe even a matplotlib Python code exporter?) as an intermediate step to any desired format would be completely sufficient. Also, if we drop any other exporter and focus on this one, we can also integrate it way more into the codebase, e.g. giving every `PlotItem` class a function like to `toMatplotlib(self, fig, ax)`. From what I can tell, this should solve most of the issues we currently have with exporters.

**Comment by 3fon3fonov** [comment]
I fully support @2xB  on this! This is exactly how I export plots at the moment; export to `matplotlib`. However, based on my experience, making the `matplotlib` GUI export reproducing exactly tor at least close the view from `pyqtgraph` is a real pain! I had to modify significantly `MatplotlibWidget.py `and `Matplotlib.py `to get it as close as possible to my pyqtgraph scatter plots. I was never able to export a TextItem, Arrows, etc. On top of that I the `matplotlib` `formlayout` is rather poor in options so I had to "monkey-jack" the `figureoptions.py` too. I can send you my modifications if you are interested, but I am afraid these may be only useful for the [Exo-Striker tool](https://github.com/3fon3fonov/exostriker).

I think the best would be to have a png/jpg/pdf export that can reproduce the pyqtgraph scene 1:1. 

p.s. my problems with the .png and the .svg exporters are explained in https://groups.google.com/g/pyqtgraph/c/EnHt19ogGh0.
a new git pull of the master version of pyqtgraph 0.11 does not fix the problems I had ~1 year ago.

**Comment by 2xB** [comment]
Yes, agreed: As a user, one should be able to export to other formats without worrying about `matplotlib`. I also agree that there is room for improvement for the current `matplotlib` exporter. For using it internally as conversion step, I so far made the experience that `matplotlib` allows quite a lot of customization, especially using style sheets. @3fon3fonov, can you summarize what styling you had issues with?

**Comment by markotoplak** [comment]
In Orange we do pyqtgraph pdf export already, here is the code: https://github.com/biolab/orange-widget-base/blob/master/orangewidget/utils/PDFExporter.py

We would be happy if this code would go upstream. It worked perfectly well with pyqtgraph 0.10.0, but was broken with  pyqtgraph/pyqtgraph#321. I am trying to find out a way to solve it for 0.11/master in biolab/orange-widget-base#111

**Comment by j9ac9k** [comment]
hey @markotoplak Always good to hear from the orange team, we're very interested when we introduce regressions, especially as we've merged a whole lot of PRs in the last year.  I'm not seeing anything obvious there, but if you do figure it out, let me know and I'll try and sneak in the fix before we release 0.11.1 which were pretty close to doing

**Comment by j9ac9k** [comment]
Also regarding the PDF exporter, I'm not opposed to merging that into the pyqtgraph library, but it would be done after we deprecate Qt4/Qt5 < 5.12 (aka when we release 0.12 (or 1.0?)

**Comment by markotoplak** [comment]
Another option is to only load it for Qt>5.12. That is what we did.

**Comment by 3fon3fonov** [comment]
Dear @markotoplak, this is excellent news! I hope this or an equivalent solution is included in pyqtgraph. Indeed, the exporter could be loaded only if PyQt5 > 5.12 is available.  

TO answer @2xB, one could in principle convert the .png --> .pdf via e.g. 

`$ convert input.png output.pdf`

However, the .png exporter also has problems. If one increases the resolution the lines get really thin, so export to .pdf does not work. 

See e.g resolution of 500x343.:

![test_500x343](https://user-images.githubusercontent.com/44244057/100391534-08dbfe80-3034-11eb-9cbc-2e9a34cc47e2.png)

and with 4000x2750

 
![test_4000x2750](https://user-images.githubusercontent.com/44244057/100391787-e991a100-3034-11eb-819b-a2e6c9d96dbc.png)



The .svg exporter works, but for some reason, the model line looks horrible, errorbars are not in the middle (or data are shifted) I get ticks also on the right and top
![test](https://user-images.githubusercontent.com/44244057/100391750-be0eb680-3034-11eb-9657-ce32c72aa34e.jpg)
:

**Comment by j9ac9k** [comment]
oh, there is a PR for a fix for the svg exporter I believe, can you try
applying the change in PR #1341 and see if that makes the SVG exporter work?

**Comment by j9ac9k** [comment]
> Another option is to only load it for Qt>5.12. That is what we did.

Definitely a legit thing to do; but as we are on the verge of deprecating so many versions of Qt, I'd rather not add more if checks on Qt versions unless really necessary, which a PDF exporter wouldn't qualify for that IMO; especially as so many issues w/ our exporter remain currently.

**Comment by 2xB** [comment]
@3fon3fonov Yes, for the image exporters, they all have their flaws. That's why I think it would be useful to only have one image exporter to worry about (e.g. matplotlib) that is able to take care about everything else.

Apart from that, a PNG exporter does not export vector graphics, which (like e.g. SVG files) describe shapes like lines and circles instead of describing pixels. So it would probably be best to keep everything in the vector graphics space as long as possible, which would be lost when using PNG.

PS: Actually, the other exporters are also not perfect, we both have open issues about bugs in the CSV exporter. But that's a different story...

**Comment by j9ac9k** [comment]
+1-ing the effort to use the SVG exporter that @2xB suggested; but @3fon3fonov I want you to try the fix in #1341 and see if that fixes the SVG output for your use case (either that or if you can provide the code to generate the above plots, that would be great too).

**Comment by 3fon3fonov** [comment]
@j9ac9k, first, I must say that using the new master code of pyqtgraph 0.11, I see some improvements in the svg exporter! For example, the right and top ticks are gone, and I see more options available (yet not really useful at the moment see below). 

Adding the three lines of code in #1341 did not help. I even wonder where is the "offset" variable used afterward? 

This is a screenshot from  my laptop:

![test_svg](https://user-images.githubusercontent.com/44244057/100433948-17f69700-309c-11eb-87ed-4d5ea16642a2.png)

As you can see I still have problems with data/errors and the model which appears like from an 8-bit console.

Playing with the width and height parameters led to this: (2000 x 1380), which I am not sure if is useful...

![test_svg_hi_res](https://user-images.githubusercontent.com/44244057/100434396-b7b42500-309c-11eb-8dc5-1d725dd72e92.png)


So if you can fix the svg import show a smooth model and data, I would be very happy to use it I even had a small script that converts the .svg plots to .pdf, although I never included it in the Exo-Striker, because of my issues with the .svg exporter.
```

#!/bin/bash

for i in $@; do
  inkscape --export-area-drawing --without-gui --export-pdf="$(basename $i .svg).pdf" $i
done

```

**Comment by j9ac9k** [comment]
Not long ago we merged https://github.com/pyqtgraph/pyqtgraph/pull/1401 which had some fixes for the SVG exporter

Can you generate a short example that highlights issues w/ the SVG exporter?  Usually working from the in the examples folder is easier here.

**Comment by j9ac9k** [comment]
@3fon3fonov We just make some changes to `ScatterPlotItem.py` via #1420 would you mind re-checking the exporter?

**Comment by 3fon3fonov** [comment]
Hi all, 

I am not sure what you changed, but even with the new version I still have the same problems as described above.

**Comment by j9ac9k** [comment]
wait, which problem...sorry we've gone a few rounds here, I'm talking about scatter plot entities should now be positioned correctly on the exporter (that is if you run pyqtgraph from the master branch), you saying that's not the case for your example?

**Comment by 3fon3fonov** [comment]
This is a fresh export from the SVG exporter.

![image](https://user-images.githubusercontent.com/44244057/102662430-1ee66600-417f-11eb-9e52-f40e97de4ef8.png)

The data and errorbars seem OK. The model is jagged as before, I got an extra y-axis ticks on the right, and height and width, do not do anything than producing the same plot but with huge white space margin.

**Comment by j9ac9k** [comment]
ok, thanks for the clarification; relative to your previous picture, the scatter plot items seem to be centered much better within the vertical line ... I was mostly commenting on the position of the scatter items... clearly there is more work to be done w/ the exporter itself.

**Comment by 3fon3fonov** [comment]
Hi again. 

I am back on this, since the .svg exporter now seems to work fine! Many thanks for the SVG exporter! 

I was thinking if it is possible to re-use the SVGExporter to also outputs a .pdf if requested (e.g., as a boolean in the "FileDialog"). For instance,  using the "cairosvg" library.

```
import cairosvg

def convert_svg_to_pdf(svg_file, pdf_file):
    try:
        cairosvg.svg2pdf(url=svg_file, write_to=pdf_file)
        print("Conversion completed successfully!")
    except Exception as e:
        print("Conversion failed:", str(e))
 
```

I understand that including additional dependencies is something pyqtgraph should avoid, but a .pdf export is something really needed. Perhaps it can be included as an optional feature. For instance, if "cairosvg" is installed by the user, then the QCheckBox will be active, else will be grayed out.
![screenshot](https://github.com/pyqtgraph/pyqtgraph/assets/44244057/a6b627a7-acd2-49d0-9cb5-539177f0cf38)

**Comment by j9ac9k** [comment]
I'm not opposed to adding a PDF exporter, but I am reluctant to add another dependency just to do so when it looks as if Qt has that functionality already per this comment:

https://github.com/pyqtgraph/pyqtgraph/issues/1455#issuecomment-734299674

Also nothing preventing you from subclassing the SVG exporter and adding that method yourself if that's a dependency you can count on yourself.

**Comment by 3fon3fonov** [comment]
Yes, OK, I did it on my own implementing an internal version of "cairoSVG" and for Linux users "rsvg-convert", which adopt the .svg file from the SVGExporter and can produce .svg, and/or .pdf/.ps. 

![image](https://github.com/pyqtgraph/pyqtgraph/assets/44244057/4b1afd5b-1b1d-4452-9b5a-b7c441786a62)

There are some problems with the fonts of "cairoSVG", whereas "rsvg-convert" will work only if installed, but I am happy!!!

 I also did attempt to include a PDFExporter from https://github.com/pyqtgraph/pyqtgraph/issues/1455#issuecomment-734299674 , but with a limited success, unfortunately. Ideally, similar PDFExporter.py code should be included, but for me this is a workaround!

Thanks!

**Comment by 2xB** [comment]
Wasn't the conclusion we had at the beginning of this discussion, that all exporters inherently have flaws and therefore it would be best for graphics exporters to just have one Matplotlib exporter that can then save to different file types? (compare https://github.com/pyqtgraph/pyqtgraph/issues/1455#issuecomment-732445422 and below) This now seems to be lost in the discussion? By that one would also avoid adding a new dependency.
